utils/markdown/renderer.tsx

```diff
@@ -95,9 +95,30 @@ export default class Renderer extends marked.Renderer {
             }
         }
 
+        const onClipboardClick = (e: any, code: string) => {
+            navigator.clipboard.writeText(code);
+
+            const classes = e.target.getAttribute('class');
+            if (classes.indexOf('icon-content-copy') > -1) {
+                e.target.setAttribute('class', classes.replace('icon-content-copy', '') + 'icon-check');
+                e.target.setAttribute('title', 'Copied');
+
+                setTimeout(() => {
+                    const classes = e.target.getAttribute('class');
+                    e.target.setAttribute('class', classes.replace('icon-check', '') + 'icon-content-copy');
+                    e.target.setAttribute('title', 'Copy code block');
+                }, 5000);
+            }
+        };
+
         return (
             '<div class="' + className + '">' +
-                header +
+                '<div class="post-code__overlay">' +
+                    '<span class="post-code__clipboard" onclick="' + this.makeOnClickFunction(onClipboardClick, code) + '">' +
```

Comment:
I appreciate that everything's sanitized here and appears to be done properly, but perhaps we can use html-to-react differently to do this as React instead of messing around with HTML more. I forgot we actually still rendered code blocks as raw HTML instead of what we do for some other parts of Markdown where the HTML is just a placeholder that gets replaced with smarter components using html-to-react. If you look at [`utils/message_html_to_component.jsx`](https://github.com/mattermost/mattermost-webapp/blob/master/utils/message_html_to_component.jsx), you'll see where we replace the simple HTML versions of things like at-mentions and emojis with React components that can do more work there.

Doing that would definitely let React do more of the lifting as you said. We'll probably still have to pass the contents of the code block through in the HTML rendered here, but it'll definitely be easier overall. At the very least, you'll probably want to do that for the copy button, but you could also move the whole code block to be a proper React component if that ends up being easier for you.

Reply:
awesome, thanks! I'll have a look probably next week

Reply:
@hmhealey can you have a look? I moved the code to a component as you mentioned. I did not have time to also implement the context menu but it will be a breeze from now :). Will do this in the following days.

Two things:
- didn't understand this part `(!('markdown' in options) || options.markdown)` isn't it basically if(true) ? Is it there only for readability? (Well I added this part but I used the same style as above)
- I have not added the copy button to the tex/latex code block since it renders a different component. There's also texcode/latexcode so I wasn't sure the former is still used. Should I also update that?

Reply:
Whoops. I forgot to respond to these questions before. I don't know if you still needed answers to these, but for curiosity's sake:
- `!('markdown' in options) || options.markdown)` will be false if `options.markdown` is specifically set to `false`, but it'll be true if `options.markdown` hasn't been defined to begin with.
- The `tex`/`latex` languages are special since they actually render the Latex inside them instead of rendering as a code block, and then we added `texcode`/`latexcode` to let people render Latex as code. As it is, I think the current code correctly adds a copy button for Latex code blocks, but not for the rendered Latex which is what I'd expect